### PR TITLE
feat(api): search workflow handler

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13216.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13241.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13241.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13246.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13252.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✘  
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13253.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13254.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✘  
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13254.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13263.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13265.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13265.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13266.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13267.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: sdk/notif.go [hunk 1]
```diff
@@ -44,6 +48,24 @@ type UserNotificationSettings struct {
 	Conditions   WorkflowNodeConditions    `json:"conditions,omitempty" yaml:"conditions,omitempty"`
 }
 
+// Value returns driver.Value from Metadata.
+func (a UserNotificationSettings) Value() (driver.Value, error) {
+	j, err := json.Marshal(a)
+	return j, WrapError(err, "cannot marshal UserNotificationSettings")
+}
+
+// Scan Metadata.
```
Comment (yesnault) [inline sdk/notif.go:57]:
```suggestion
// Scan UserNotificationSettings.
```

File: engine/api/workflow/dao.go [hunk 2]
```diff
@@ -10,82 +10,47 @@ import (
...
-	return getAll(ctx, db, query)
+func LoadAllNamesByProjectIDs(ctx context.Context, db gorp.SqlExecutor, projectIDs []int64) ([]sdk.WorkflowName, error) {
+	query := `
+    SELECT workflow.*, project.projectkey
+	FROM workflow
+	JOIN project ON project.id = workflow.project_id
+    WHERE project_id = ANY($1)`
+
+	var result []sdk.WorkflowName // This struct is not registered as a gorpmapping entity so we can't use gorpmapping.Query
+	_, err := db.Select(&result, query, pq.Int64Array(projectIDs))
```
Comment (richardlt) [inline engine/api/workflow/dao.go:38]: Catch ErrNoRows to returns empty list ?

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13287.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/api/ascode/dao.go [hunk 3]
```diff
@@ -28,6 +30,21 @@ func LoadEventByWorkflowIDAndPullRequest(ctx context.Context, db gorp.SqlExecuto
 	return &event, nil
 }
 
+func LoadEventsByWorkflowIDs(ctx context.Context, db gorp.SqlExecutor, workflowIDs []int64) ([]sdk.AsCodeEvent, error) {
+	// ?| => Do any of these array strings exist as top-level keys?
```
Comment (richardlt) [inline engine/api/ascode/dao.go:34]:
```suggestion
```

File: engine/api/pipeline/pipeline.go [hunk 4]
```diff
@@ -239,6 +240,34 @@ func DeletePipeline(ctx context.Context, db gorp.SqlExecutor, pipelineID int64)
...
+// LoadAllByIDs loads all pipelines
+func LoadAllByIDs(db gorp.SqlExecutor, ids []int64, loadDependencies bool) ([]sdk.Pipeline, error) {
+	var pips []sdk.Pipeline
+	query := `SELECT id, name, description, project_id, last_modified, from_repository
+			  FROM pipeline
+			  WHERE id = ANY($1)
+			  ORDER BY pipeline.name`
+
+	if _, err := db.Select(&pips, query, pq.Int64Array(ids)); err != nil {
+		return nil, sdk.WithStack(err)
```
Comment (richardlt) [inline engine/api/pipeline/pipeline.go:252]: Check no rows

File: engine/api/workflow.go [hunk 5]
```diff
@@ -605,26 +598,36 @@ func (api *API) deleteWorkflowHandler() service.Handler {
 		consumer := getAPIConsumer(ctx)
 		sdk.GoRoutine(api.Router.Background, "deleteWorkflowHandler",
 			func(ctx context.Context) {
-				txg, errT := api.mustDB().Begin()
-				if errT != nil {
-					log.Error(ctx, "deleteWorkflowHandler> Cannot start transaction: %v", errT)
+				txg, err := api.mustDB().Begin()
+				if err != nil {
+					log.Error(ctx, "deleteWorkflowHandler> Cannot start transaction: %v", err)
```
Comment (richardlt) [inline engine/api/workflow.go:603]: return

File: engine/api/workflow_ascode.go [hunk 6]
```diff
@@ -105,7 +105,7 @@ func (api *API) postWorkflowAsCodeHandler() service.Handler {
 		if err := workflow.RenameNode(ctx, api.mustDB(), &wf); err != nil {
 			return err
 		}
-		if err := workflow.IsValid(ctx, api.Cache, api.mustDB(), &wf, *p, workflow.LoadOptions{DeepPipeline: true}); err != nil {
+		if err := workflow.CompleteWorkflow(ctx, api.mustDB(), &wf, *p, workflow.LoadOptions{DeepPipeline: true}); err != nil {
```
Comment (richardlt) [inline engine/api/workflow_ascode.go:108]: I think here we also need to check validity because the workflow comes from the request body.

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13293.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Pull request merged**